origin_ui/origin_api.go

```diff
@@ -0,0 +1,189 @@
+		directorTimeoutTimer = time.NewTimer(directorTimeoutDuration)
+		go func() {
+			for {
+				select {
+				case <-directorTimeoutTimer.C:
+					// Timer fired because no message was received in time.
+					log.Warningln("No director test report received within the time limit")
+					if err := metrics.SetComponentHealthStatus("director", "critical", "No director test report received within the time limit"); err != nil {
+						log.Errorln("Failed to update director component health status:", err)
+					}
+					// Reset the timer for the next period.
+					timerMutex.Lock()
+					directorTimeoutTimer.Reset(directorTimeoutDuration)
+					timerMutex.Unlock()
+				case <-exitLoop:
+					log.Infoln("Gracefully terminating the director-health test timeout loop...")
+					return
+				}
+			}
+		}()
+	} else {
+		if !directorTimeoutTimer.Stop() {
+			<-directorTimeoutTimer.C
+		}
+		directorTimeoutTimer.Reset(directorTimeoutDuration)
+	}
+}
+
+// Director will periodiclly upload/download files to/from all connected
+// origins and test the health status of origins. It will send a request
+// reporting such status to this endpoint, and we will update origin internal
+// health status metric to reflect the director connection status.
+func directorTestResponse(ctx *gin.Context) {
+	dt := DirectorTest{}
+	if err := ctx.ShouldBind(&dt); err != nil {
+		log.Errorf("Invalid director test response")
+		ctx.JSON(400, gin.H{"error": "Invalid director test response"})
+		return
+	}
+	// We will let the timer go timeout if director didn't send a valid json request
+	resetDirectorTimeoutTimer()
+	if dt.Status == "ok" {
+		if err := metrics.SetComponentHealthStatus("director", "ok", fmt.Sprintf("Director timestamp: %s", dt.Timestamp)); err != nil {
+			log.Errorln("Failed to update director component health status:", err)
+			ctx.JSON(500, gin.H{"error": fmt.Sprintf("Failed to update director component health status: %s", err)})
+			return
+		}
+	} else if dt.Status == "error" {
+		if err := metrics.SetComponentHealthStatus("director", "critical", dt.Message); err != nil {
+			log.Errorln("Failed to update director component health status:", err)
+			ctx.JSON(500, gin.H{"error": fmt.Sprintf("Failed to update director component health status: %s", err)})
+			return
+		}
+	} else {
+		log.Errorf("Invalid director test response, status: %s", dt.Status)
+		ctx.JSON(400, gin.H{"error": fmt.Sprintf("Invalid director test response status: %s", dt.Status)})
+	}
+}
+
+func ConfigureOriginAPI(router *gin.Engine) error {
+	if router == nil {
+		return errors.New("Origin configuration passed a nil pointer")
+	}
+	if err := metrics.SetComponentHealthStatus("director", "warning", "Initializing origin, unknown status for director"); err != nil {
+		return err
+	}
+	// start the timer for the director test report timeout
+	resetDirectorTimeoutTimer()
+
+	// When program exits
+	signal.Notify(exitSignals, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
```

Comment:
As follow-up work, it's probably better to provide `ConfigureOriginAPI` a context object that can be cancelled on shutdown.

Having signal handlers scattered throughout the code is a bit scary -- too much indirect action -- and are fairly difficult to unit test compared to a context.